polygeist/lib/Dialect/Polygeist/Transforms/DetectReduction.cpp

```diff
@@ -73,35 +73,174 @@ class ReductionOp {
   return OS;
 }
 
-class AffineForReductionIter : public OpRewritePattern<AffineForOp> {
+class LoopReductionIter {
+public:
+  LoopReductionIter(const LoopReductionIter &) = delete;
+  LoopReductionIter(LoopReductionIter &&) = delete;
+  LoopReductionIter &operator=(const LoopReductionIter &) = delete;
+  LoopReductionIter &operator=(LoopReductionIter &&) = delete;
+
+protected:
   Pass::Statistic &numReductionsDetected;
 
-public:
-  AffineForReductionIter(MLIRContext *context,
-                         Pass::Statistic &numReductionsDetected)
-      : OpRewritePattern<AffineForOp>(context),
-        numReductionsDetected(numReductionsDetected) {}
+  LoopReductionIter(Pass::Statistic &numReductionsDetected)
+      : numReductionsDetected(numReductionsDetected) {}
+  virtual ~LoopReductionIter() = default;
+
+  virtual Block::BlockArgListType
+  getRegionIterArgs(LoopLikeOpInterface Loop) const = 0;
 
+  virtual unsigned getNumRegionIterArgs(LoopLikeOpInterface Loop) const = 0;
+
+  virtual void cloneFilteredTerminator(
+      Operation *Terminator,
+      const SmallVectorImpl<ReductionOp> &ReductionOps) const = 0;
+
+  virtual LoopLikeOpInterface
+  createNewLoop(LoopLikeOpInterface Loop, SmallVectorImpl<Value> &NewIterArgs,
+                PatternRewriter &Rewriter) const = 0;
+
+  LogicalResult matchAndRewrite(LoopLikeOpInterface Loop,
+                                PatternRewriter &Rewriter) const {
+    SmallVector<ReductionOp> ReductionOps;
+    WalkResult Result = collectCandidates(Loop, ReductionOps);
+    if (Result.wasInterrupted() || ReductionOps.empty()) {
+      LLVM_DEBUG({
+        llvm::dbgs() << "No reduction found\n";
+        llvm::dbgs() << "------------------------------------------------\n";
+      });
+      return failure();
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "\nFound " << ReductionOps.size() << " reduction(s):\n";
+      for (ReductionOp &Op : ReductionOps)
+        llvm::dbgs() << Op << "\n";
+      llvm::dbgs() << "\n";
+    });
+
+    // Move the load outside the loop (recall that the load is loop invariant).
+    // The load result is passed to the new loop as an iter argument.
+    SmallVector<Value, 4> NewIterArgs(getRegionIterArgs(Loop));
+    Rewriter.setInsertionPoint(Loop);
+    LLVM_DEBUG(llvm::dbgs() << "New Loop:\n");
+    for (const ReductionOp &Op : ReductionOps) {
+      Operation *MovedLoad = Rewriter.clone(*Op.getLoad());
+      LLVM_DEBUG(llvm::dbgs() << *MovedLoad << "\n");
```

Comment:
```suggestion
      LLVM_DEBUG(llvm::dbgs().indent(2) << *MovedLoad << "\n");
```

Reply:
The indent isn't necessary because what is printed are load instructions hoisted out of the loop. Currently the transformed loop is printed as (focus on the first 2 load instructions):
```
New Loop:
%0 = affine.load %alloca_1[0] : memref<1xf32>
%1 = affine.load %alloca[0] : memref<1xf32>
%9:2 = "scf.for"(%2, %4, %3, %7, %8) ({
^bb0(%arg1: index, %arg2: f32, %arg3: f32):
  %13 = "memref.load"(%arg0, %arg1) : (memref<?xf32>, index) -> f32
  %14 = "arith.addf"(%arg3, %13) {fastmath = #arith.fastmath<none>} : (f32, f32) -> f32
  %15 = "arith.mulf"(%arg2, %13) {fastmath = #arith.fastmath<none>} : (f32, f32) -> f32
  "scf.yield"(%15, %14) : (f32, f32) -> ()
}) : (index, index, index, f32, f32) -> (f32, f32)
"affine.store"(%9#0, %6) {map = affine_map<() -> (0)>} : (f32, memref<1xf32>) -> ()
"affine.store"(%9#1, %5) {map = affine_map<() -> (0)>} : (f32, memref<1xf32>) -> ()
------------------------------------------------
```

Reply:
ok